heartbeat/monitors/resolver.go

```diff
@@ -0,0 +1,94 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitors
+
+import (
+	"fmt"
+	"net"
+)
+
+// Resolver lets us define custom DNS resolvers similar to what the go stdlib provides, but
+// potentially with custom functionality
+type Resolver interface {
+	// ResolveIPAddr is an analog of net.ResolveIPAddr
+	ResolveIPAddr(network string, host string) (*net.IPAddr, error)
+	// LookupIP is an analog of net.LookupIP
+	LookupIP(host string) ([]net.IP, error)
+}
+
+// stdResolver uses the go std library to perform DNS resolution.
+type stdResolver struct{}
+
+// We only ever need one instance of this
+var StdResolver = stdResolver{}
+
+func (s stdResolver) ResolveIPAddr(network string, host string) (*net.IPAddr, error) {
+	return net.ResolveIPAddr(network, host)
+}
+
+func (s stdResolver) LookupIP(host string) ([]net.IP, error) {
+	return net.LookupIP(host)
+}
+
+// staticResolver allows for a custom in-memory mapping of hosts to IPs, it ignores network names
+// and zones.
+type staticResolver struct {
+	mapping map[string][]net.IP
+}
+
+func CreateStaticResolver() staticResolver {
```

Comment:
Do not return private types from an exported function/method. Why do you need CreateStaticResolver to be exported? If it is only for testing, how about having a special package with testing utilities, to make the intent more clear?

When introducing something like this resolver (so we don't have to modify the /etc/hosts file), then the resolver should become 'stackable', so we can fallback to another resolver if the hostname has not been configured.

If we use this for testing, how about accepting a `map[string][]net.IP` as parameter to `CreateStaticResolver`?

Reply:
It doesn't need to be exported, just an oversight :)

Reply:
I take it back, it does need to be exported because it's used by the tcp monitor package tests. I could move it there, but I think it makes more sense to keep it here since we'll eventually use it in tests for http/icmp.

Reply:
Alternatively I could extract all this resolver stuff to its own package, but I'm indifferent on that matter.

Reply:
If it's purpose is to be used for testing only, then we should at least have this reflected in the package name.

Reply:
Well, `StdResolver` is used for production, `StaticResolver` is used in another package for test purposes. I could break it out to its own package, but that seems a bit excessive, since that package would have to export it.

We could put `StaticResolver` in a test file in the `tcp` package, but in the future we'll probably have tests that want to reuse it from the `http` and `icmp` monitors.

What do you think the best way forward is?

Reply:
Can we think of other testing helpers? E.g. some helper that would drive a 'job' with it's followup tasks?

Reply:
This has been moved to the tcp test package for now.